modyn/evaluator/internal/grpc/evaluator_grpc_servicer.py

```diff
@@ -0,0 +1,223 @@
+                "batches_seen": num_batches,
+                "samples_seen": num_samples,
+            }
+            cleaned_kwargs = {k: v for k, v in response_kwargs_running.items() if v is not None}
+            return EvaluationStatusResponse(**cleaned_kwargs)  # type: ignore[arg-type]
+
+        exception = self._check_for_evaluation_exception(evaluation_id)
+        logger.info(
+            f"Evaluation {evaluation_id} is no longer running. "
+            f"Process finished {'successfully' if not exception else 'with errors'}."
+        )
+        response_kwargs_finished: dict[str, Any] = {
+            "valid": True,
+            "is_running": False,
+            "blocked": False,
+            "state_available": False,
+            "exception": exception,
+        }
+        cleaned_kwargs = {k: v for k, v in response_kwargs_finished.items() if v is not None}
+        return EvaluationStatusResponse(**cleaned_kwargs)  # type: ignore[arg-type]
+
+    def _get_status(self, evaluation_id: int) -> tuple[Optional[int], Optional[int]]:
+        status_query_queue = self._evaluation_process_dict[evaluation_id].status_query_queue
+        status_query_queue.put(EvaluatorMessages.STATUS_QUERY_MESSAGE)
+        try:
+            # blocks for 30 seconds
+            response = self._evaluation_process_dict[evaluation_id].status_response_queue.get(timeout=30)
+            return response["num_batches"], response["num_samples"]
+        except queue.Empty:
+            return None, None
+
+    def _check_for_evaluation_exception(self, evaluation_id: int) -> Optional[str]:
+        exception_queue = self._evaluation_process_dict[evaluation_id].exception_queue
+
+        # As qsize() is unreliable and not implemented on macOS,
+        # we try to fetch an element within 100ms. If there is no
+        # element within that timeframe returned, we return None.
+        try:
+            exception = exception_queue.get(timeout=0.1)
+            return exception
+        except queue.Empty:
+            return None
+
+    def get_final_evaluation(
+        self, request: FinalEvaluationRequest, context: grpc.ServicerContext
+    ) -> FinalEvaluationResponse:
+        evaluation_id = request.evaluation_id
+        logger.info(f"Received get final evaluation request for evaluation {evaluation_id}.")
+
+        if evaluation_id not in self._evaluation_dict:
+            logger.error(f"Evaluation with id {evaluation_id} has not been registered.")
+            return FinalEvaluationResponse(valid=False)
+
+        if self._evaluation_process_dict[evaluation_id].process_handler.is_alive():
+            logger.error(f"Evaluation with id {evaluation_id} is still running.")
+            return FinalEvaluationResponse(valid=False)
+
+        logger.error("Returning results of all metrics.")
+        evaluation_data: list[EvaluationData] = []
+        for metric in self._evaluation_dict[evaluation_id].metrics:
```

Comment:
as discussed during our call, the dict is not shared, so we cannot access the results like that

Reply:
Agree, see comment above :)